pkg/reconciler/channel/channel.go

```diff
@@ -0,0 +1,412 @@
+				return err
+			}
+			c.Recorder.Eventf(channel, corev1.EventTypeNormal, "UpdatedSubscriber", "Updated Subscriber %q", ps.Name)
+		}
+		for i, ss := range channel.Status.Subscribers {
+			if ss.UID == s.UID {
+				channel.Status.Subscribers[i].ObservedGeneration = s.Generation
+			}
+		}
+		return nil
+	}
+	for _, s := range subDeletes {
+		genName := resources.GenerateSubscriptionName(s.UID)
+		c.Logger.Infof("Channel %q will delete subscription %s", channel.Name, genName)
+		// TODO: we need to handle the case of a already deleted pull subscription. Perhaps move to ensure deleted method.
+		if err := c.RunClientSet.PubsubV1alpha1().PullSubscriptions(channel.Namespace).Delete(genName, &metav1.DeleteOptions{}); err != nil {
+			c.Logger.Errorf("unable to delete PullSubscription %s for Channel %q, %s", genName, channel.Name, err.Error())
+			c.Recorder.Eventf(channel, corev1.EventTypeWarning, "DeleteSubscriberFailed", "Deleting Subscriber %q failed", genName)
+			return err
+		}
+		c.Recorder.Eventf(channel, corev1.EventTypeNormal, "DeletedSubscriber", "Deleted Subscriber %q", genName)
+
+		for i, ss := range channel.Status.Subscribers {
+			if ss.UID == s.UID {
+				// Swap len-1 with i and then pop len-1 off the slice.
+				channel.Status.Subscribers[i] = channel.Status.Subscribers[len(channel.Status.Subscribers)-1]
+				channel.Status.Subscribers = channel.Status.Subscribers[:len(channel.Status.Subscribers)-1]
+			}
+		}
+		return nil // Signal a re-reconcile.
+	}
+
+	return nil
+}
+
+func (r *Reconciler) createTopic(ctx context.Context, channel *v1alpha1.Channel) (*pubsubv1alpha1.Topic, error) {
+	topic, err := r.getTopic(ctx, channel)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logging.FromContext(ctx).Error("Unable to get an Topic", zap.Error(err))
+		return nil, err
+	}
+	if topic != nil {
+		logging.FromContext(ctx).Desugar().Info("Reusing existing Topic", zap.Any("topic", topic))
+		if topic.Status.Address != nil {
+			channel.Status.SetAddress(topic.Status.Address.URL)
+		} else {
+			channel.Status.SetAddress(nil)
+		}
+		return topic, nil
+	}
+	topic, err = r.RunClientSet.PubsubV1alpha1().Topics(channel.Namespace).Create(resources.MakeTopic(&resources.TopicArgs{
+		Owner:   channel,
+		Name:    resources.GenerateTopicName(channel.UID),
+		Project: channel.Spec.Project,
+		Secret:  channel.Spec.Secret,
+		Topic:   channel.Status.TopicID,
+		Labels:  resources.GetLabels(controllerAgentName, channel.Name),
+	}))
+	if err != nil {
+		logging.FromContext(ctx).Info("Topic created.", zap.Error(err), zap.Any("topic", topic))
+		r.Recorder.Eventf(channel, corev1.EventTypeNormal, "TopicCreated", "Created Topic %q", topic.GetName())
+	}
+	return topic, err
+}
+
+func (r *Reconciler) getTopic(ctx context.Context, channel *v1alpha1.Channel) (*pubsubv1alpha1.Topic, error) {
+	tl, err := r.RunClientSet.PubsubV1alpha1().Topics(channel.Namespace).List(metav1.ListOptions{
```

Comment:
Since you don't use generateName, you can just do a get instead of listing things?

Reply:
I now get, not list. Thanks.